File: flowcatalyst-typescript/apps/flowcatalyst/src/platform/domain/identity-provider/events.ts
/**
 * Identity Provider Domain Events
 */

import {
	BaseDomainEvent,
	DomainEvent,
	type ExecutionContext,
} from "@flowcatalyst/domain-core";

const APP = "platform";
const DOMAIN = "iam";
const SOURCE = `${APP}:${DOMAIN}`;

// -----------------------------------------------------------------------------
// IdentityProviderCreated
// -----------------------------------------------------------------------------

export interface IdentityProviderCreatedData {
	readonly identityProviderId: string;
	readonly code: string;
	readonly name: string;
	readonly type: string;
	readonly [key: string]: unknown;
}

export class IdentityProviderCreated extends BaseDomainEvent<IdentityProviderCreatedData> {
	static readonly EVENT_TYPE = DomainEvent.eventType(
		APP,
		DOMAIN,
		"identity-provider",
		"created",
	);
	static readonly SPEC_VERSION = "1.0";

	constructor(ctx: ExecutionContext, data: IdentityProviderCreatedData) {
		super(
			{
				eventType: IdentityProviderCreated.EVENT_TYPE,
				specVersion: IdentityProviderCreated.SPEC_VERSION,
				source: SOURCE,
				subject: DomainEvent.subject(
					APP,
					"identity-provider",
					data.identityProviderId,
				),
				messageGroup: DomainEvent.messageGroup(
					APP,
					"identity-provider",
					data.identityProviderId,
				),
			},
			ctx,
			data,
		);
	}
}

// -----------------------------------------------------------------------------
// IdentityProviderUpdated
// -----------------------------------------------------------------------------

export interface IdentityProviderUpdatedData {
	readonly identityProviderId: string;
	readonly name: string;
	readonly type: string;
	readonly [key: string]: unknown;
}

export class IdentityProviderUpdated extends BaseDomainEvent<IdentityProviderUpdatedData> {
	static readonly EVENT_TYPE = DomainEvent.eventType(
		APP,
		DOMAIN,
		"identity-provider",
		"updated",
	);
	static readonly SPEC_VERSION = "1.0";

	constructor(ctx: ExecutionContext, data: IdentityProviderUpdatedData) {
		super(
			{
				eventType: IdentityProviderUpdated.EVENT_TYPE,
				specVersion: IdentityProviderUpdated.SPEC_VERSION,
				source: SOURCE,
				subject: DomainEvent.subject(
					APP,
					"identity-provider",
					data.identityProviderId,
				),
				messageGroup: DomainEvent.messageGroup(
					APP,
					"identity-provider",
					data.identityProviderId,
				),
			},
			ctx,
			data,
		);
	}
}

// -----------------------------------------------------------------------------
// IdentityProviderDeleted
// -----------------------------------------------------------------------------

export interface IdentityProviderDeletedData {
	readonly identityProviderId: string;
	readonly code: string;
	readonly [key: string]: unknown;
}

export class IdentityProviderDeleted extends BaseDomainEvent<IdentityProviderDeletedData> {
	static readonly EVENT_TYPE = DomainEvent.eventType(
		APP,
		DOMAIN,
		"identity-provider",
		"deleted",
	);
	static readonly SPEC_VERSION = "1.0";

	constructor(ctx: ExecutionContext, data: IdentityProviderDeletedData) {
		super(
			{
				eventType: IdentityProviderDeleted.EVENT_TYPE,
				specVersion: IdentityProviderDeleted.SPEC_VERSION,
				source: SOURCE,
				subject: DomainEvent.subject(
					APP,
					"identity-provider",
					data.identityProviderId,
				),
				messageGroup: DomainEvent.messageGroup(
					APP,
					"identity-provider",
					data.identityProviderId,
				),
			},
			ctx,
			data,
		);
	}
}

File: internal/platform/identityprovider/operations/events.go
package operations

import (
	"encoding/json"
	"time"

	"github.com/flowcatalyst/flowcatalyst-go/pkg/fcsdk/usecase"
)

const (
	IdentityProviderCreatedType = "platform:admin:identity-provider:created"
	IdentityProviderUpdatedType = "platform:admin:identity-provider:updated"
	IdentityProviderDeletedType = "platform:admin:identity-provider:deleted"
	Source                      = "platform:admin"
)

func subjectFor(id string) string { return "platform.identityprovider." + id }
func groupFor(id string) string   { return "platform:identityprovider:" + id }

type IdentityProviderCreated struct {
	Metadata           usecase.EventMetadata
	IdentityProviderID string
	Code               string
}

func (e IdentityProviderCreated) EventID() string       { return e.Metadata.EventID }
func (e IdentityProviderCreated) EventType() string     { return IdentityProviderCreatedType }
func (e IdentityProviderCreated) SpecVersion() string   { return "1.0" }
func (e IdentityProviderCreated) Source() string        { return Source }
func (e IdentityProviderCreated) Subject() string       { return subjectFor(e.IdentityProviderID) }
func (e IdentityProviderCreated) Time() time.Time       { return e.Metadata.OccurredAt }
func (e IdentityProviderCreated) PrincipalID() string   { return e.Metadata.PrincipalID }
func (e IdentityProviderCreated) CorrelationID() string { return e.Metadata.CorrelationID }
func (e IdentityProviderCreated) CausationID() string   { return e.Metadata.CausationID }
func (e IdentityProviderCreated) ExecutionID() string   { return e.Metadata.ExecutionID }
func (e IdentityProviderCreated) MessageGroup() string  { return groupFor(e.IdentityProviderID) }
func (e IdentityProviderCreated) ToDataJSON() ([]byte, error) {
	return json.Marshal(struct {
		IdentityProviderID string `json:"identityProviderId"`
		Code               string `json:"code"`
	}{e.IdentityProviderID, e.Code})
}

type IdentityProviderUpdated struct {
	Metadata           usecase.EventMetadata
	IdentityProviderID string
	Code               string
}

func (e IdentityProviderUpdated) EventID() string       { return e.Metadata.EventID }
func (e IdentityProviderUpdated) EventType() string     { return IdentityProviderUpdatedType }
func (e IdentityProviderUpdated) SpecVersion() string   { return "1.0" }
func (e IdentityProviderUpdated) Source() string        { return Source }
func (e IdentityProviderUpdated) Subject() string       { return subjectFor(e.IdentityProviderID) }
func (e IdentityProviderUpdated) Time() time.Time       { return e.Metadata.OccurredAt }
func (e IdentityProviderUpdated) PrincipalID() string   { return e.Metadata.PrincipalID }
func (e IdentityProviderUpdated) CorrelationID() string { return e.Metadata.CorrelationID }
func (e IdentityProviderUpdated) CausationID() string   { return e.Metadata.CausationID }
func (e IdentityProviderUpdated) ExecutionID() string   { return e.Metadata.ExecutionID }
func (e IdentityProviderUpdated) MessageGroup() string  { return groupFor(e.IdentityProviderID) }
func (e IdentityProviderUpdated) ToDataJSON() ([]byte, error) {
	return json.Marshal(struct {
		IdentityProviderID string `json:"identityProviderId"`
		Code               string `json:"code"`
	}{e.IdentityProviderID, e.Code})
}

type IdentityProviderDeleted struct {
	Metadata           usecase.EventMetadata
	IdentityProviderID string
	Code               string
}

func (e IdentityProviderDeleted) EventID() string       { return e.Metadata.EventID }
func (e IdentityProviderDeleted) EventType() string     { return IdentityProviderDeletedType }
func (e IdentityProviderDeleted) SpecVersion() string   { return "1.0" }
func (e IdentityProviderDeleted) Source() string        { return Source }
func (e IdentityProviderDeleted) Subject() string       { return subjectFor(e.IdentityProviderID) }
func (e IdentityProviderDeleted) Time() time.Time       { return e.Metadata.OccurredAt }
func (e IdentityProviderDeleted) PrincipalID() string   { return e.Metadata.PrincipalID }
func (e IdentityProviderDeleted) CorrelationID() string { return e.Metadata.CorrelationID }
func (e IdentityProviderDeleted) CausationID() string   { return e.Metadata.CausationID }
func (e IdentityProviderDeleted) ExecutionID() string   { return e.Metadata.ExecutionID }
func (e IdentityProviderDeleted) MessageGroup() string  { return groupFor(e.IdentityProviderID) }
func (e IdentityProviderDeleted) ToDataJSON() ([]byte, error) {
	return json.Marshal(struct {
		IdentityProviderID string `json:"identityProviderId"`
		Code               string `json:"code"`
	}{e.IdentityProviderID, e.Code})
}
